graspy/embed/oosase.py

```diff
@@ -0,0 +1,416 @@
+    directed : bool
+        True if the graph is directed
+    id_to_index : dict
+        Mapping from node id to node index (i.e. 'node_id0' -> 0)
+    index_to_id : dict
+        Mapping from node index to node id (i.e. 0 -> 'node_id0')
+
+
+
+    See Also
+    --------
+    graspy.embed.selectSVD
+    graspy.embed.select_dimension
+
+    Notes
+    -----
+    The singular value decomposition: 
+
+    .. math:: A = U \Sigma V^T
+
+    is used to find an orthonormal basis for a matrix, which in our case is the adjacency
+    matrix of the graph. These basis vectors (in the matrices U or V) are ordered according 
+    to the amount of variance they explain in the original matrix. By selecting a subset of these
+    basis vectors (through our choice of dimensionality reduction) we can find a lower dimensional 
+    space in which to represent the graph.
+
+    References
+    ----------
+    .. [1] Sussman, D.L., Tang, M., Fishkind, D.E., Priebe, C.E.  "A
+       Consistent Adjacency Spectral Embedding for Stochastic Blockmodel Graphs,"
+       Journal of the American Statistical Association, Vol. 107(499), 2012
+    .. [2] Levin, K., Roosta-Khorasani, F., Mahoney, M. W., & Priebe, C. E. (2018). Out-of-sample 
+        extension of graph adjacency spectral embedding. PMLR: Proceedings of Machine Learning 
+        Research, 80, 2975-2984.
+    """
+
+    def __init__(
+        self,
+        n_components=None,
+        n_elbows=2,
+        algorithm="randomized",
+        n_iter=5,
+        check_lcc=True,
+        in_sample_proportion=1,
+        in_sample_idx=None,
+        connected_attempts=100,
+        semi_supervised=False,
+        random_state=None,
+    ):
+        super().__init__(
+            n_components=n_components,
+            n_elbows=n_elbows,
+            algorithm=algorithm,
+            n_iter=n_iter,
+            check_lcc=check_lcc,
+        )
+
+        if random_state is None:
+            random_state = np.random.randint(10 ** 6)
+        np.random.seed(random_state)
+
+        if in_sample_proportion <= 0 or in_sample_proportion > 1:
+            if in_sample_idx is None:
+                msg = (
+                    "must give either proportion of in sample indices or a list"
+                    + " of in sample vertices"
+                )
+                raise ValueError(msg)
+        self.connected_attempts = connected_attempts
+        self.semi_supervised = semi_supervised
+        self.in_sample_proportion = in_sample_proportion
+        self.in_sample_idx = in_sample_idx
+
+    def fit(self, graph, y=None):
+        """
+        Fit ASE model to input graph
+
+        Parameters
+        ----------
+        graph : array_like or networkx.Graph
+            Input graph to embed.
+
+        Returns
+        -------
+        self : returns an instance of self.
+        """
+
+        if isinstance(graph, np.ndarray):
+            check_array(graph)
+            graph = nx.Graph(graph).copy()
```

Comment:
have you tested this with directed data? I believe both of the simulated data tests you did will only yield undirected graphs. I am guessing this line will break (or give you the wrong thing secretly) because you are forcing it to be an undirected nx.Graph(). Also I don't think you need the `.copy()` here?

Reply:
no i have not.